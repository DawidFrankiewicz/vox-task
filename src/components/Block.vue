<script setup lang="ts">
import { ref } from 'vue';

defineProps<{ id: string; backgroundColor: string }>();
const emit = defineEmits<{
    (e: 'removeBlock', id: string): void;
    (e: 'editBlock', id: string, newColorValue: string): void;
    (e: 'swapPositions', id: string, targetId: string): void;
}>();

const colorPicker = ref<HTMLInputElement | null>(null);

const isValidHexColorValue = (stringToTest: string): boolean => {
    // Hex color regexes for 3, 4, 6 and 8 digit hex values, ignoring case
    const hex3ColorRegex = /^#[0-9a-f]{3}$/i;
    const hex4ColorRegex = /^#[0-9a-f]{4}$/i;
    const hex6ColorRegex = /^#[0-9a-f]{6}$/i;
    const hex8ColorRegex = /^#[0-9a-f]{8}$/i;

    const regexes: RegExp[] = [
        hex6ColorRegex,
        hex3ColorRegex,
        hex8ColorRegex,
        hex4ColorRegex,
    ];

    // If any of the regexes matches, return true
    for (const regex of regexes) {
        if (regex.test(stringToTest)) {
            return true;
        }
    }

    return false;
};

const editBlock = (id: string, newHexColorValue: string): void => {
    if (isValidHexColorValue(newHexColorValue)) {
        emit('editBlock', id, newHexColorValue);
    }
};

const onDragStart = (event: DragEvent, id: string): void => {
    if (event.dataTransfer) {
        event.dataTransfer.dropEffect = 'move';
        event.dataTransfer.effectAllowed = 'move';
        event.dataTransfer.setData('id', id);
    }
};

const onDrop = (event: DragEvent, targetId: string): void => {
    const id = event.dataTransfer?.getData('id');

    if (id) {
        // Prevent event when dropping on itself
        if (id === targetId) {
            return;
        }

        emit('swapPositions', id, targetId);
    }
};
</script>

<template>
    <div
        data-test-id="block"
        class="block"
        :style="`background-color: ${$props.backgroundColor}`"
        draggable="true"
        @dragstart="onDragStart($event, $props.id)"
        @drop="onDrop($event, $props.id)"
        @dragenter.prevent
        @dragover.prevent
    >
        <button
            data-test-id="removeBlockButton"
            @click="emit('removeBlock', $props.id)"
        >
            Usuń
        </button>
        <label class="block__color-picker button">
            Zmień
            <input
                data-test-id="colorPickerInput"
                ref="colorPicker"
                type="color"
                :value="$props.backgroundColor"
                @input="editBlock($props.id, String(colorPicker?.value))"
            />
        </label>
    </div>
</template>

<style lang="scss" scoped>
.block {
    display: flex;
    flex-wrap: wrap;
    justify-content: space-between;
    align-items: flex-end;
    align-content: flex-end;
    gap: 5px;
    padding: 1em;
    width: 100%;
    height: 250px;
    // Default background color if not provided in props or invalid hex value
    background-color: #000;

    @media screen and (min-width: 340px) {
        width: 250px;
    }

    &__color-picker {
        position: relative;

        input[type='color'] {
            visibility: hidden;
            pointer-events: none;
            position: absolute;
            left: 50%;
            top: 50%;
            transform: translate(-50%, -50%);
        }
    }
}
</style>
