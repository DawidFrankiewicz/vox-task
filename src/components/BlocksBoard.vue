<script setup lang="ts">
import { Block } from '../types';

import { ref } from 'vue';
import BlockDisplay from './BlockDisplay.vue';
import { v4 as uuidv4 } from 'uuid';

const blocks = ref<Block[]>([]);

const randomRgbColorValue = (): [number, number, number] => {
    // Get random RGB values (0-255)
    const randomR = Math.round(Math.random() * 256);
    const randomG = Math.round(Math.random() * 256);
    const randomB = Math.round(Math.random() * 256);

    return [randomR, randomG, randomB];
};

const randomHexColorValue = (): string => {
    const [randomR, randomG, randomB] = randomRgbColorValue();

    // Convert RGB values to hex, pad with 0 if needed
    const hexR = randomR.toString(16).padStart(2, '0');
    const hexG = randomG.toString(16).padStart(2, '0');
    const hexB = randomB.toString(16).padStart(2, '0');

    // Return hex color value in format #RRGGBB
    return `#${hexR}${hexG}${hexB}`;
};

const addRandomBlock = (): void => {
    const newBlock: Block = {
        // Generate random UUID
        id: uuidv4(),
        color: randomHexColorValue(),
    };
    blocks.value.push(newBlock);
};

const removeBlock = (id: string): void => {
    blocks.value = blocks.value.filter((block) => block.id !== id);
};

const editBlock = (id: string, newColorValue: string): void => {
    const foundBlock = blocks.value.find((block) => block.id === id);
    if (foundBlock) {
        foundBlock.color = newColorValue;
    }
};
</script>

<template>
    <div class="board__container">
        <h1 class="board__header">VOX Kolorowe Kafelki</h1>

        <div class="controlls__container">
            <button @click="addRandomBlock">Dodaj Kafelek</button>
        </div>

        <div class="blocks__container">
            <BlockDisplay
                v-for="block in blocks"
                :key="block.id"
                @removeBlock="removeBlock(block.id)"
                @editBlock="
                    (newColorValue) => editBlock(block.id, newColorValue)
                "
                :backgroundColor="block.color"
            />
        </div>
    </div>
</template>

<style lang="scss" scoped>
.board__container {
    padding: 2em clamp(0.5em, 5vw, 3em);
    display: flex;
    flex-direction: column;
    align-items: center;
}

.board__header {
    text-align: center;
    font-size: clamp(1.5em, 5vw, 3em);
}

.controlls__container {
    margin-bottom: 1em;
}
.blocks__container {
    display: flex;
    flex-wrap: wrap;
    justify-content: center;
    align-items: center;
    gap: 10px;
}
</style>
