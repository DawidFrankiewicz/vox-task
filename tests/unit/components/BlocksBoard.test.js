import { mount } from '@vue/test-utils';
import Block from '@/components/Block.vue';
import BlocksBoard from '@/components/BlocksBoard.vue';

const factory = async () => {
    return mount(BlocksBoard, {
        shallow: true,
    });
};

const deepFactory = async () => {
    return mount(BlocksBoard);
};

test('Should render correctly', async () => {
    const wrapper = await factory();

    expect(wrapper.find('[data-test-id="addRandomBlockButton"]').exists()).toBe(
        true
    );

    // Check if component matches snapshot (HTML structure)
    // https://vitest.dev/guide/snapshot#updating-snapshots
    expect(wrapper.html()).toMatchSnapshot();
});

test('Should add block with random color', async () => {
    const wrapper = await factory();

    // Hex color regex
    const hex6ColorRegex = /^#[0-9A-F]{6}$/i;

    const button = wrapper.find('[data-test-id="addRandomBlockButton"]');
    await button.trigger('click');
    await wrapper.vm.$nextTick();

    const block = wrapper.find('[data-test-id="block"]');
    expect(block.exists()).toBe(true);
    expect(block.attributes('backgroundcolor')).toMatch(hex6ColorRegex);
});

test('Should remove block', async () => {
    const wrapper = await deepFactory();

    // Add random block
    const button = wrapper.find('[data-test-id="addRandomBlockButton"]');
    await button.trigger('click');

    const block = wrapper.findComponent(Block);
    expect(block.exists()).toBe(true);

    // Emit removeBlock event from Block component
    block.vm.$emit('removeBlock', block.vm.id, block.vm.backgroundColor);
    await wrapper.vm.$nextTick();

    expect(wrapper.find('[data-test-id="block"]').exists()).toBe(false);
});

test('Should edit block', async () => {
    const wrapper = await deepFactory();

    // Add random block
    const button = wrapper.find('[data-test-id="addRandomBlockButton"]');
    await button.trigger('click');

    const block = wrapper.findComponent(Block);
    expect(block.exists()).toBe(true);

    // Emit editBlock event from Block component
    block.vm.$emit('editBlock', block.vm.id, '#000000');
    await wrapper.vm.$nextTick();

    expect(block.vm.backgroundColor).toBe('#000000');
});
