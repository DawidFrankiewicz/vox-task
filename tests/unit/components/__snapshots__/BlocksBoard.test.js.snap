// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Should render correctly 1`] = `
"<div data-v-10abbbab="" class="board__container">
  <h1 data-v-10abbbab="" class="board__header">VOX Kolorowe Kafelki</h1>
  <div data-v-10abbbab="" class="controlls__container"><button data-v-10abbbab="" data-test-id="addRandomBlockButton"> Dodaj Kafelek </button></div>
  <div data-v-10abbbab="" class="blocks__container"></div>
</div>"
`;
